examples/pennylane_run_rotoselect.py

```diff
@@ -0,0 +1,405 @@
+    params[d] = np.pi / 2.
+    M_0_plus = cost(params, generators)
+    params[d] = -np.pi / 2.
+    M_0_minus = cost(params, generators)
+    a = np.arctan2(2. * M_0 - M_0_plus - M_0_minus, M_0_plus - M_0_minus)
+    params[d] = -np.pi / 2. - a
+    if params[d] > np.pi:
+        params[d] -= 2 * np.pi
+    if params[d] <= -np.pi:
+        params[d] += 2 * np.pi
+    return cost(params, generators)
+
+def optimal_theta_and_gen_helper(d, params, generators, cost):
+    params[d] = 0.
+    M_0 = cost(params, generators)  # M_0 independent of generator selection
+    for generator in ['X', 'Y', 'Z']:
+        generators[d] = generator
+        params_cost = rotosolve_d(d, params, generators, cost, M_0)
+        if generator == 'X' or params_cost <= params_opt_cost:
+            params_opt_d = params[d]
+            params_opt_cost = params_cost
+            generators_opt_d = generator
+    return params_opt_d, generators_opt_d
+
+
+def rotoselect_cycle(cost, params, generators):
+    for d in range(len(params)):
+        params[d], generators[d] = optimal_theta_and_gen_helper(d, params, generators, cost)
+    return params, generators
+
+
+##############################################################################
+# Optimizing the circuit structure
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+# We perform the optimization and print the optimal generators for the rotation gates. The cost function is
+# reduced below the minimal value from gradient descent or Rotosolve on the original
+# circuit structure ansatz. In other words, the learned circuit structure performs better without
+# increasing the depth of the circuit.
+
+costs_rotoselect = []
+params = init_params[:]
+init_generators = ['X', 'Y']
+generators = init_generators
+for _ in range(n_steps):
+    costs_rotoselect.append(cost(params, generators))
+    params, generators = rotoselect_cycle(cost, params, generators)
+
+print("Optimal generators are: {}".format(generators))
+
+# plot cost function vs. steps comparison
+fig, (ax1,ax2) = plt.subplots(1,2,figsize=(7,3))
+plt.subplot(1,2,1)
+plt.plot(steps,costs_grad_desc,'o-')
+plt.title("grad. desc. on original ansatz")
+plt.xlabel("steps")
+plt.ylabel("cost")
+plt.subplot(1,2,2)
+plt.plot(steps,costs_rotoselect,'o-')
+plt.title("rotoselect")
+plt.xlabel("cycles")
+plt.ylabel("cost")
+plt.yticks(np.arange(-1.25,0.80,0.25))
+plt.tight_layout()
+plt.show()
+
+
+##############################################################################
+# Cost function surface for learned circuit structure
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+#
+# .. figure:: ../../examples/figures/learned_structure.png
+#    :scale: 65%
+#    :align: center
+#    :alt: learned_structure
+#
+# |
+# Finally, we plot the cost function surface for the newly discovered optimized
+# circuit structure shown in the figure above. It is apparent from the minima in the plot that
+# the original ansatz was not expressive enough to arrive at the optimal values
```

Comment:
I don't think "expressive" is the right term to use here. The learned structure is just as expressive as the original ansatz; it just wasn't as well-matched to the problem

Reply:
Changed to "It is apparent from the minima in the plot that the new circuit structure is better suited for the problem. "